### ISSUE TITLE
CPU performance issues

### ISSUE THREAD

**Issue by raver119** [deeplearning4j/nd4j#1179]
With current nd4j/libnd4j master i'm definitely experiencing "lower then expected" performance with nd4j-native backend.

TLDR: after rc3.10 release, we had everything working with SIMD, etc, but i definitely have issues with some of ops:

For example:
- i'm unable to get reported time for broadcasts over 1024x1024 matrix, even with **parallel for** totally removed. Even in that case it's around 300k ns, instead of expected 75k-100k ns. With ***parallel for*** time gets to 1ms, as it was before rc3.10
- reduce along dimension with 1024x1024 matrix takes 22ms with setOmpNumThreads(8), and 5ms with setOmpNumThreads(1);
- same for PairwiseTransforms.

However, SIMD stuff & inlining definitely works, since on _lower_ data dimensions i get valid time for scalars and transforms, which match 3.10 values.

Overall performance on example such as LenetMNIST is also lower then expected: 
- around 90-120 seconds per epoch with setOmpNumThreads(2)
- around 220-250 seconds per epoch with setOmpNumThreads(8)

Profiler screenshots will be posted in subsequent comments.

PS: This old issue contains some numbers for 3.10 improvements: https://github.com/deeplearning4j/nd4j/issues/986

Tests were made using all-updated Ubuntu 16.04, gcc 5.4, glibc 2.23, the same PC as during rc3.10 baking: Core i7-4790, 32GB RAM.

**Comment by raver119** [comment]
Checked the same example against 0.4.0.

Results are equal to setOmpNumThreads(2): 90-130 seconds per epoch.  CPU use is around 200-240%, so it looks like 0.4.0 was compiled with strict limitation on OMP_NUM_THREADS.

P.S. batchSize is 64 during tests.

**Comment by raver119** [comment]
Overall top methods, sorted by Own Time:

![screenshot from 2016-08-02 21-31-13](https://cloud.githubusercontent.com/assets/12250879/17340402/9d4f9954-58f8-11e6-9e97-6d1041ab5cd1.png)

**Comment by raver119** [comment]
15% of all training time is spent in IsMax:

![screenshot from 2016-08-02 21-34-29](https://cloud.githubusercontent.com/assets/12250879/17340483/f5a2af4c-58f8-11e6-8199-0b79652c9f09.png)

**Comment by saudet** [comment]
I've looked at that 2 weeks ago and we're going to need to do some cache blocking to improve performance further: https://software.intel.com/en-us/articles/cache-blocking-techniques

**Comment by raver119** [comment]
I wonder what was changed, since i'm unable to reproduce my own timings received during 3.10 preparations :(

**Comment by raver119** [comment]
@AlexDBlack provided results from his ubuntu 16.04

execBroadcastOp: Execution time: 722891
execReduceOp1: Execution time: 5978855
Result: 1048576.0
execReduceOp2: Execution time: 40928
execPairwiseOp1: Execution time: 71238
testScalarOp1: Execution time: 3200
Divi result: 5001.5

**Comment by treo** [comment]
If you have a test case that you are running, I can try to see how it performs for me, didn't have much time to test anything much since the rc3.10 release

**Comment by raver119** [comment]
That would be awesome. 

There's set of tests available, they just print out time. I guess it's time to bring them to nd4j-benchmark though.

Here they are: https://github.com/deeplearning4j/nd4j/blob/0db21d49a1eab1edd9fc3fa3461384f6d291f568/nd4j-backends/nd4j-backend-impls/nd4j-native/src/test/java/org/nd4j/linalg/cpu/nativecpu/ops/NativeOpExecutionerTest.java#L22-L22

**Comment by treo** [comment]
As I'm seeing very different results on each run, I guess that speed step is kicking my butt here.... In order to fix that, I'll whip up a new JMH Suite of tests based on those tests, so we can have something that isn't as dependent on powersaving settings and too low iteration counts.

**Comment by raver119** [comment]
Yea, i'm going to unpack old benchmarks again.

**Comment by saudet** [comment]
Erratic processing time is probably caused by cache trashing.

Also, IsMax is probably calculating way more offsets than is actually required.

**Comment by AlexDBlack** [comment]
Speaking of IsMax, I still wanted to replace it with a "setMaxAlongDimension" type op. Currently we do isMax to give us essentially a one-hot vector in each row; then we do a muliColumnVector to set the value we want in the correct position.
Just mentioning it before anyone spends too much time on IsMax...

**Comment by raver119** [comment]
IsMax isn't the only issue that bothers me. Overall performance is gimped for two set of ops: broadcast and pairwise, ops that do use OpenMP threads. Like it was with rc3.9, when using single thread was giving better performance.

**Comment by treo** [comment]
So I've updated my benchmarks here: https://github.com/treo/benchmarking_nd4j
Everything was run using the following command line:
`java -jar target/benchmarks.jar NativeOps -wi 10 -i 10 -f 2`

Locally built from master, using mkl and its libsvml:
```
Benchmark                Mode     Cnt     Score    Error  Units
NativeOps.broadcastOp  sample  313523    56.457 ±  1.381  us/op
NativeOps.pairwiseOp1  sample  405427    22.157 ±  3.080  us/op
NativeOps.reduceOp1    sample    9149  2186.112 ± 29.866  us/op
NativeOps.reduceOp2    sample  284123    44.564 ±  2.761  us/op
NativeOps.scalarOp1    sample  480982     3.720 ±  5.128  us/op
NativeOps.transformOp  sample  240430    83.137 ±  1.681  us/op
```

0.5.0 Release from Maven:
```
Benchmark                Mode     Cnt     Score    Error  Units
NativeOps.broadcastOp  sample  272537    73.273 ±  1.616  us/op
NativeOps.pairwiseOp1  sample  385377    22.809 ±  4.052  us/op
NativeOps.reduceOp1    sample   12478  1602.578 ± 20.280  us/op
NativeOps.reduceOp2    sample  319272    61.285 ±  2.052  us/op
NativeOps.scalarOp1    sample  468722     3.458 ±  1.821  us/op
NativeOps.transformOp  sample  349522    57.163 ±  1.135  us/op
```

0.4.0 Release from Maven:
```
Benchmark                Mode     Cnt     Score    Error  Units
NativeOps.broadcastOp  sample  279874    71.371 ±  1.604  us/op
NativeOps.pairwiseOp1  sample  572150     4.467 ±  3.036  us/op
NativeOps.reduceOp1    sample   11613  1721.569 ± 22.967  us/op
NativeOps.reduceOp2    sample  333717    59.861 ±  1.758  us/op
NativeOps.scalarOp1    sample  513647     2.551 ±  0.025  us/op
NativeOps.transformOp  sample  293024    68.183 ±  1.439  us/op
```

0.4-rc3.10 from Maven:
```
Benchmark                Mode     Cnt     Score    Error  Units
NativeOps.broadcastOp  sample  307059    65.062 ±  1.239  us/op
NativeOps.pairwiseOp1  sample  437346     3.720 ±  1.404  us/op
NativeOps.reduceOp1    sample   11830  1690.498 ± 21.929  us/op
NativeOps.reduceOp2    sample  330053    60.785 ±  2.384  us/op
NativeOps.scalarOp1    sample  495955     3.408 ±  2.067  us/op
NativeOps.transformOp  sample  352639    56.642 ±  0.888  us/op
```

0.4-rc3.9 from Maven:
```
Benchmark                Mode     Cnt     Score    Error  Units
NativeOps.broadcastOp  sample   15198  1315.962 ± 28.055  us/op
NativeOps.pairwiseOp1  sample  422579    23.148 ±  0.871  us/op
NativeOps.reduceOp1    sample    6615  3026.184 ± 51.688  us/op
NativeOps.reduceOp2    sample   23693   844.341 ± 17.501  us/op
NativeOps.scalarOp1    sample  450234    17.025 ±  0.659  us/op
NativeOps.transformOp  sample  334298    57.285 ±  1.020  us/op
```

Given those numbers we can see pretty clearly that pairwise op has regressed to its pre rc3.10 speed.

**Comment by raver119** [comment]
Current master, glibc 2.23
```
Benchmark                Mode     Cnt    Score    Error  Units
NativeOps.broadcastOp  sample  228810   87.353 ±  2.478  us/op
NativeOps.pairwiseOp1  sample  389892   31.468 ±  1.466  us/op
NativeOps.reduceOp1    sample   35268  566.680 ± 11.640  us/op
NativeOps.reduceOp2    sample  213616   93.659 ±  4.305  us/op
NativeOps.scalarOp1    sample  465255    3.019 ±  0.007  us/op
NativeOps.transformOp  sample  234909   85.081 ±  2.009  us/op
```

**Comment by raver119** [comment]
So yea, i'm totally agree about pairwise. But also i wonder why broadcast grows for me...

**Comment by treo** [comment]
Your broadcast result doesn't look any larger than mine, given that you seem to have overall larger scores.

**Comment by raver119** [comment]
@treo could you please remind, what cpu do you have there?

**Comment by treo** [comment]
That's an i7-6700K running at 4.5GHz

**Comment by raver119** [comment]
related issue: https://github.com/deeplearning4j/nd4j/issues/1241

**Comment by raver119** [comment]
As of now: 
```
Benchmark                 Mode     Cnt    Score   Error  Units
NativeOps.broadcastOp1  sample  411473    9.019 ± 5.505  us/op
NativeOps.broadcastOp2  sample  327675   53.927 ± 0.917  us/op
NativeOps.pairwiseOp1   sample  386397   15.737 ± 0.575  us/op
NativeOps.reduceOp1     sample   46719  427.902 ± 6.992  us/op
NativeOps.reduceOp2     sample  250488   45.186 ± 2.842  us/op
NativeOps.scalarOp1     sample  468592    2.982 ± 0.005  us/op
NativeOps.transformOp   sample  270872   73.780 ± 1.868  us/op
```

**Comment by raver119** [comment]
So from my point of view issue is resolved. I've got all timings at desired levels +-

**Comment by raver119** [comment]
A bit more extended report:

```
Benchmark                         Mode     Cnt     Score      Error  Units
NativeOps.broadcastOp1          sample  413795     9.381 ±    7.578  us/op
NativeOps.broadcastOp2          sample  261173    53.510 ±    0.103  us/op
NativeOps.dupDifferentOrdersOp  sample  122128   163.584 ±    0.438  us/op
NativeOps.dupSameOrdersOp       sample  370308    23.942 ±    0.989  us/op
NativeOps.pairwiseOp1           sample  462413  3503.467 ± 1434.522  ns/op
NativeOps.reduceOp1             sample   37525   532.670 ±    4.204  us/op
NativeOps.reduceOp2             sample  339133    58.898 ±    1.423  us/op
NativeOps.scalarOp1             sample  495279     5.496 ±    7.385  us/op
NativeOps.scalarOp2             sample  387841    16.868 ±    0.859  us/op
NativeOps.transformOp           sample  283881    61.502 ±    0.959  us/op
```

TL/DR: broadcast + pairwise got improvements, transform runs slighty better now too. Scalar is tbd on small values, but on larger arrays runs way better.

**Comment by raver119** [comment]
Also, to be mentioned: HT sucks.
